tmf-transform.exs

```diff
@@ -0,0 +1,363 @@
+        end
+      )
+
+    headers
+    |> List.flatten()
+  end
+
+  # built an Artifact key by combining Rows A and B, optionally
+  # inserting a "Section" header, ie a value from row A, but in
+  # a previous column.
+  defp combine_header_rows(a, b, section)
+  defp combine_header_rows(a, "", _), do: {"", a}
+  defp combine_header_rows("", b, ""), do: {"", b}
+  defp combine_header_rows("", b, section), do: {"", concat_section_header(section, b)}
+  defp combine_header_rows(a, b, ""), do: {a, concat_section_header(a, b)}
+
+  # Build a string from 2 values
+  defp concat_section_header(section, header) do
+    Enum.join([section, header], @section_separator)
+  end
+
+  # Split string into a 2 element list
+  defp split_key_to_section_and_header(key) do
+    String.split(key, @section_separator, trim: true, parts: 2)
+  end
+
+  # Return true if the string is an output of concat_section_header/2
+  defp has_section_marker?(str), do: String.contains?(str, @section_separator)
+
+  # creates a new map within the top level artifact that corresponds
+  # to a section from one of the 2 header rows in the spreadsheet.
+  defp add_section_to_artifact(artifact, key, value) do
+    # split key into section/header
+    [section, header] = split_key_to_section_and_header(key)
+
+    {_, new_artifact} =
+      artifact
+      |> Map.get_and_update(section, fn
+        current_value when is_nil(current_value) ->
+          {current_value, Map.new([{header, value}])}
+
+        current_value ->
+          {current_value, current_value |> Map.put(header, value)}
+      end)
+
+    new_artifact
+  end
+end
+
+defmodule TmfReferenceModel.Transformer.Glossary do
+
+  def transform(package, sheet) do
+    with {:ok, rows} <- XlsxReader.sheet(package, sheet, empty_rows: false, number_type: String)
+    do
+      {:ok, rows |> transform_rows()}
+    end
+  end
+
+  def transform_rows(rows) do
+    rows
+    |> Enum.reject(fn [key | _] ->
+      key
+      |> String.trim()
+      |> case do
+        "Abbreviation" -> true
+        "Zone" -> true
+        "Item" -> true
+        _  -> false
+      end
+    end)
+    |> Enum.map_reduce([], fn row, acc ->
+      row
+      |> parse()
+      |> case do
+       {:merged, definition} ->
+          # The spreadsheet may use 2 cells to define an item.
+          # when this happens, the "key" value is empty, but there is still a "value".
+          # parse/1 will notice this, and return a :merged atom
+          # to tell this function to look back into the accumulator for the last item
+          # added, and replace that value.
+
+          #TODO: make this special case handling cleaner
+          [{key, value} | _] = acc
+
+          {nil, acc |> List.replace_at(0, {key, value <> " " <> definition})}
+
+        any -> {nil, [any | acc]}
+      end
+    end)
+    |> then(fn {_, list} -> list end)
+    |> List.flatten()
+    |> Enum.into(%{})
+  end
```

Comment:
That's some sweet mixology right there.

(Mixology being my newly coined term for the act of writing Elixir code.)